integration_test.go

```diff
@@ -0,0 +1,231 @@
+
+	srv := NewServer("tcp", "localhost:"+IntegrationServerPort, mux1)
+	srv.Debug = true
+
+	// This is used to signal back to the calling test that the server has started up
+	fmt.Println("PID:", os.Getpid())
+	if err := srv.ListenAndServe(); err != nil {
+		fmt.Println(err)
+	}
+}
+
+// handler used by the integration test server
+func integrationTestServerHandler(w http.ResponseWriter, r *http.Request) {
+	duration, err := time.ParseDuration(r.FormValue("duration"))
+	if err != nil {
+		http.Error(w, err.Error(), 400)
+		return
+	}
+	time.Sleep(duration)
+	fmt.Fprintf(
+		w, "Slept for %2v seconds. Response from pid %d.\n",
+		duration.Seconds(),
+		os.Getpid(),
+	)
+}
+
+// createIntegrationServer will re-run the test binary that is currently running,
+// but only execute the "TestIntegrationServer" test, which will stand up an
+// 'endless' server which can be used to test signal handling
+func createIntegrationServerTest() *exec.Cmd {
+	cmd := exec.Command(os.Args[0], "-test.run=TestIntegrationServer")
+
+	// Set the feature toggle for the integration test server
+	cmd.Env = []string{IntegrationServerFeatureToggle + "=1"}
+
+	return cmd
+}
+
+func TestIntegrationRestart(t *testing.T) {
+
+	intServer := createIntegrationServerTest()
+	serverMessages := make(chan string) // channel to receive notifications from the integration server
+	// Start test integration server
+	require.NoError(t, startIntegrationServer(intServer, serverMessages), "starting integration test server")
+	parentPid := intServer.Process.Pid
+	defer teardownServer(parentPid)
+	go intServer.Wait() // when this ends the parent process should be over and pipe clean up will happen
+
+	// The pid is printed out from the integation test server once it has fully started up
+	// Check for this message so we can start sending requests/signals to the server
+	msg, msgChannelOpen := <-serverMessages
+	require.True(t, msgChannelOpen, "Unexpected close of server messages channel")
+	msgPid, _ := strconv.Atoi(msg)
+	require.Equal(t, parentPid, msgPid, "Unexpected message down server messages channel")
+	time.Sleep(50 * time.Millisecond) // quick sleep just to give the server a chance to bind to port
+
+	// Send a long running request to the test server
+	longReqWait := &sync.WaitGroup{}
+	longReqWait.Add(1)
+	var (
+		longResp     *http.Response
+		longReqError error
+	)
+	go func() {
+		longResp, longReqError = sendIntegrationTestRequest(2 * time.Second)
+		longReqWait.Done()
+	}()
+
+	time.Sleep(50 * time.Millisecond) // quick sleep just to give the request a chance to get handled
+
+	// Send a signal to the test server to restart
+	require.NoError(t, intServer.Process.Signal(syscall.SIGHUP), "send signal to restart integration test server")
+
+	// The child server should also print it's pid out once it's started up, so look for this
+	msg, msgChannelOpen = <-serverMessages
+	require.True(t, msgChannelOpen, "Unexpected close of server messages channel")
+	msgPid, _ = strconv.Atoi(msg)
```

Comment:
lost err

Reply:
This is an unusual case where there was a comment that explained why, which I then deleted as, in my warped sense of appropriate comments, I thought it was redundant. I don't mind throwing away the error here as:
* We've already checked it's an int in the function that sends messages down the channel
* If there was an error anyway, msgPid will be 0, which the next line then checks it isn't, or in the other place we do this, we immediately compare it to the pid of the integration server, where it won't be equal

I can add a comment to explain